content/blog/2024/anchor-position-yearbook.md

```diff
@@ -0,0 +1,163 @@
+row. It's not a terribly difficult layout for print, but there are a few
+challenges to implement it in HTML.
+
+First, it would be a challenge to have the name and the photo of each person
+together in the HTML. Also, if it needed to be responsive, with different
+numbers of photos per row, I'd likely have to have multiple layouts defined in
+CSS, and potentially duplicate the items in the HTML, and conditionally show
+them based on some magic number.
+
+When working with anchor positioning, I've found it helpful to borrow the
+concept of constraints from mobile development, where you describe your layout
+in terms of relationships between elements. So, let's start by defining our
+constraints-
+
+1. The left side of the names should be `1em` to the right of the photos.
+2. Each name should be in the same row as the photo, starting at the top.
+3. Each name should be below the name before it.
+
+This is all possible with anchor positioning, and we can even have the image and
+name inside of the same `<li>`.
+
+```html
+<ul>
+  <li>
+    <img src="photo1.jpg" />
+    <h2>Name 1</h2>
+  </li>
+  <li>
+    <img src="photo2.jpg" />
+    <h2>Name 2</h2>
+  </li>
+  ...
+</ul>
+```
+
+We'll make the `<ul>` the main photo area, use `flex-wrap: wrap` to fit as many
+photos in a row as possible, and make it an anchor by giving it the name
+`--photos`.
+
+Then we can place the `<h2>` titles to meet the first constraint.
+
+```css
+h2{
+  position: absolute;
+  left: calc( anchor(--photos right) + 1em );
+}
+```
+
+That's right- we don't need to first figure out how many pixels from the left
+viewport edge the photos are, we can just declaratively describe what we want.
+"Place the left edge at the sum of photos' right side and 1em".
+
+We can then satisfy the second contraint, that each name should be in the same
+row as its photo.
+
+```css
+li{
+  anchor-scope: --photo;
+}
+h2{
+  top: anchor(--photo top);
+}
+```
+
+The `top` rule simply aligns the top of the name with the top of the photo. The
+`anchor-scope` is important here. By default, if there are multiple elements
+that have the same `anchor-name`, the last element in the DOM will be selected.
+Setting `anchor-scope` an a parent element means that the anchored element won't
+look outside of the parent element for an anchor. This means the `<h2>` will be
+aligned with the top of its photo.
+
+But wait- we have one more constraint. At this point, all the names in a row are
+layered on top of each other. We want them to vertically stack instead.
+
+```css
+h2{
+  anchor-name: --title;
+  top: max( anchor(--photo top), anchor(--title bottom, 0));
+}
+```
+
+Now, the top of the name will be either the top of the photo OR the bottom of
+the title above it, which ever is more (or further down the page). That's
+because the anchoring element has to come **before** the anchored element. That
+means the reference to `anchor(--title)` will be the name above it.
```

Comment:
```suggestion
Now, the top of the name will be either the top of the photo OR the bottom of
the title above it, whichever is greater (or further down the page). That's
because the anchoring element has to come **before** the anchored element, which
means the reference to `anchor(--title)` will be the name above it.
```